executor/analyze.go

```diff
@@ -2280,3 +2281,74 @@ func analyzePKIncremental(colExec *analyzePKIncrementalExec) *statistics.Analyze
 		Snapshot: colExec.snapshot,
 	}
 }
+
+// addNewAnalyzeJob records the new analyze job.
+func addNewAnalyzeJob(ctx sessionctx.Context, job *statistics.AnalyzeJob) {
+	if job == nil {
+		return
+	}
+	statsHandle := domain.GetDomain(ctx).StatsHandle()
+	err := statsHandle.RecordAnalyzeJob(job, ctx.GetSessionVars().ConnectionID)
+	if err != nil {
+		ctx.GetSessionVars().StmtCtx.AppendWarning(err)
+	}
+}
+
+// startAnalyzeJob marks the state of the analyze job as running and sets the start time.
+func startAnalyzeJob(ctx sessionctx.Context, job *statistics.AnalyzeJob) {
+	if job == nil {
+		return
+	}
+	job.StartTime = time.Now()
+	exec := ctx.(sqlexec.RestrictedSQLExecutor)
+	const sql = "UPDATE mysql.analyze_jobs SET start_time = CONVERT_TZ(%?, '+00:00', @@TIME_ZONE), state = %? WHERE id = %?"
+	_, _, err := exec.ExecRestrictedSQL(context.TODO(), nil, sql, job.StartTime.UTC().Format(types.TimeFormat), statistics.AnalyzeRunning, job.ID)
+	if err != nil {
+		ctx.GetSessionVars().StmtCtx.AppendWarning(err)
+	}
+}
+
+// updateAnalyzeJob updates count of the processed rows when increment reaches a threshold.
+func updateAnalyzeJob(ctx sessionctx.Context, job *statistics.AnalyzeJob, rowCount int64) {
+	const maxDelta = 1000000
+	if job == nil {
+		return
+	}
+	var delta int64
+	job.Delta.Lock()
+	job.Delta.Count += rowCount
+	if job.Delta.Count > maxDelta {
+		delta = job.Delta.Count
+		job.Delta.Count = 0
+	}
+	job.Delta.Unlock()
+	if delta == 0 {
+		return
+	}
+	exec := ctx.(sqlexec.RestrictedSQLExecutor)
+	const sql = "UPDATE mysql.analyze_jobs SET processed_rows = processed_rows + %? WHERE id = %?"
+	_, _, err := exec.ExecRestrictedSQL(context.TODO(), nil, sql, delta, job.ID)
+	if err != nil {
+		ctx.GetSessionVars().StmtCtx.AppendWarning(err)
+	}
+}
+
+// finishAnalyzeJob updates the state of the analyze job to finished/failed according to `meetError` and sets the end time.
+func finishAnalyzeJob(ctx sessionctx.Context, job *statistics.AnalyzeJob, meetError bool) {
+	if job == nil {
+		return
+	}
+	job.EndTime = time.Now()
+	var state string
+	if meetError {
+		state = statistics.AnalyzeFailed
+	} else {
+		state = statistics.AnalyzeFinished
+	}
+	exec := ctx.(sqlexec.RestrictedSQLExecutor)
+	const sql = "UPDATE mysql.analyze_jobs SET end_time = CONVERT_TZ(%?, '+00:00', @@TIME_ZONE), state = %? WHERE id = %?"
```

Comment:
I'm not familiar with timezone, what's the display result if inserted this way?
If we do it in UTC+8, the result would display as the UTC+0?

Reply:
Timestamp is stored as UTC+0 in disk and the query result is displayed according to `@@TIME_ZONE`. If the system timezone is the same as `@@TIME_ZONE` of the session, we can just insert the current timestamp under system timezone without timezone conversion. The session that executes the sql is pulled from session pool and I'm not sure whether its `@@TIME_ZONE` is modified. Hence we convert UTC+0 time to `@@TIME_ZONE` to ensure that we save the right UTC+0 time in disk.